### PR TITLE
CLOUDP-95679: Added support for pem entry in tls secret

### PR DESCRIPTION
**Pull request by bznein**
This PR adds the possibility of providing directly a tls.pem entry. 
Existing secrets with tls.crt and tls.key will continue to work, this possibility is already present in enterprise and will bridge the two implementations.

There is also unit and e2e testing.

### All Submissions:

* [ ] Have you opened an Issue before filing this PR?
* [ ] Have you signed our [CLA](https://www.mongodb.com/legal/contributor-agreement)?
* [ ] Have you signed all of your commits?
* [ ] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?
* [ ] Put `closes #XXXX` in your comment to auto-close the issue that your PR fixes (if such).

### CODE REVIEW

File: docs/RELEASE_NOTES.md [hunk 1]
```diff
@@ -4,7 +4,8 @@
 
 - Changes
   - MongoDB database of the statefulSet is managed using distinct Role, ServiceAccount and RoleBinding.
-
+  - TLS Secret can also contain a single "tls.pem" entry, containing the concatenation of the certificate and key
+    - If a TLS secret contains all three entries, the operator will raise an error if the tls.pem entry is not equal to the concatenation of the tls.crt and tls.key entries
```
Comment (irajdeep) [inline docs/RELEASE_NOTES.md:8]: > If a TLS secret contains all three entries

this is a concatenation of `crt` and `key` right? so three ?

File: api/v1/mongodbcommunity_types.go [hunk 2]
```diff
@@ -321,6 +321,8 @@ type TLS struct {
 	// CertificateKeySecret is a reference to a Secret containing a private key and certificate to use for TLS.
 	// The key and cert are expected to be PEM encoded and available at "tls.key" and "tls.crt".
 	// This is the same format used for the standard "kubernetes.io/tls" Secret type, but no specific type is required.
+	// Alternatively, an entry tls.pem, containing the concatenation of cert and key, can be provided.
+	// If all three entries are provided, the tls.pem one needs to be equal to the concatenation fo tls.crt and tls.key
```
Comment (irajdeep) [inline api/v1/mongodbcommunity_types.go:325]:
```suggestion
	// If all three entries are provided, the tls.pem one needs to be equal to the concatenation of tls.crt and tls.key
```

File: controllers/mongodb_tls.go [hunk 3]
```diff
@@ -90,7 +87,7 @@ func getTLSConfigModification(getUpdateCreator secret.GetUpdateCreator, mdb mdbv
 		return automationconfig.NOOP(), nil
 	}
 
-	certKey, err := getCertAndKey(getUpdateCreator, mdb)
+	certKey, err := getPemOrConcatCrtAndKey(getUpdateCreator, mdb)
```
Comment (irajdeep) [inline controllers/mongodb_tls.go:90]: I think the name `getPemOrConcatCrtAndKey` is a bit confusing , maybe we can remove the `OrConcatCrtAndKey` since it is an implementation details of the function ?

File: controllers/mongodb_tls.go [hunk 4]
```diff
@@ -99,18 +96,27 @@ func getTLSConfigModification(getUpdateCreator secret.GetUpdateCreator, mdb mdbv
 }
 
 // getCertAndKey will fetch the certificate and key from the user-provided Secret.
-func getCertAndKey(getter secret.Getter, mdb mdbv1.MongoDBCommunity) (string, error) {
+func getCertAndKey(getter secret.Getter, mdb mdbv1.MongoDBCommunity) string {
 	cert, err := secret.ReadKey(getter, tlsSecretCertName, mdb.TLSSecretNamespacedName())
 	if err != nil {
-		return "", err
+		return ""
```
Comment (irajdeep) [inline controllers/mongodb_tls.go:102]: 👍

File: test/e2e/setup/setup.go [hunk 5]
```diff
@@ -56,7 +61,7 @@ func Setup(t *testing.T) *e2eutil.Context {
 
 // CreateTLSResources will setup the CA ConfigMap and cert-key Secret necessary for TLS
 // The certificates and keys are stored in testdata/tls
-func CreateTLSResources(namespace string, ctx *e2eutil.Context) error { //nolint
+func CreateTLSResources(namespace string, ctx *e2eutil.Context, secretType tlsSecretType) error { //nolint
```
Comment (irajdeep) [inline test/e2e/setup/setup.go:64]: now related to this PR, curious why do we have a `//nolint` here ?

**Review by irajdeep — COMMENTED** [review]
LGTM overall, added some comments inline.

File: controllers/mongodb_tls_test.go [hunk 6]
```diff
@@ -215,29 +224,89 @@ func TestCombineCertificateAndKey(t *testing.T) {
...
-	err := c.Create(context.TODO(), &s)
-	if err != nil {
-		return err
-	}
+		r := NewReconciler(client.NewManagerWithClient(c))
+
+		err = r.ensureTLSResources(mdb)
+		assert.NoError(t, err)
+
+		// Operator-managed secret should have been created and contain the
```
Comment (irajdeep) [inline controllers/mongodb_tls_test.go:241]:
```suggestion
		// Operator-managed secret should have been created and contains the
```

File: test/e2e/setup/setup.go [hunk 7]
```diff
@@ -77,22 +82,29 @@ func CreateTLSResources(namespace string, ctx *e2eutil.Context) error { //nolint
...
+		cert, err := ioutil.ReadFile(path.Join(testDataDir, "server.crt"))
+		if err != nil {
+			return err
+		}
+		key, err := ioutil.ReadFile(path.Join(testDataDir, "server.key"))
+		if err != nil {
+			return err
+		}
+		certKeySecretBuilder.SetField("tls.crt", string(cert)).SetField("tls.key", string(key))
+	} else if secretType == Pem {
```
Comment (irajdeep) [inline test/e2e/setup/setup.go:100]: do we need an `else if` instead of just `if` , since the secretType can be only one ?

File: docs/RELEASE_NOTES.md (hunk 1 above)
Comment (bznein) [reply docs/RELEASE_NOTES.md:8]: I don't get the question😅 We both wrote "three"

Comment (irajdeep) [reply docs/RELEASE_NOTES.md:8]: of sorry I mean `two` : -)

File: controllers/mongodb_tls.go (hunk 3 above)
Comment (bznein) [reply controllers/mongodb_tls.go:90]: I called it `getFinalPem` earlier but I didn't like it. We already have `getCertAndKey` and `getPem` so I wanted a name that was clearly different.

File: test/e2e/setup/setup.go (hunk 5 above)
Comment (bznein) [reply test/e2e/setup/setup.go:64]: good question :) next commit I'll try to remove it and see what happens!

File: controllers/mongodb_tls.go (hunk 3 above)
Comment (irajdeep) [reply controllers/mongodb_tls.go:90]: I see in that case maybe s/Concat/Concatenated? since the current signature, it looks like the function is performing concatenation -- will leave it upto you

File: docs/RELEASE_NOTES.md (hunk 1 above)
Comment (bznein) [reply docs/RELEASE_NOTES.md:8]: I think three is correct: the error happens if the secret contains:

- tls.key
- tls.crt
- tls.pem

and pem is different from key+crt.

Do you think I should reword it to be more clear?

File: test/e2e/setup/setup.go (hunk 5 above)
Comment (bznein) [reply test/e2e/setup/setup.go:64]: golanci-lint does not raise any error, removed!

File: docs/RELEASE_NOTES.md (hunk 1 above)
Comment (irajdeep) [reply docs/RELEASE_NOTES.md:8]: ah got it 
> Do you think I should reword it to be more clear?

yes, please I think it's mentioned at multiple places , we can make it a bit more clearer

Comment (bznein) [reply docs/RELEASE_NOTES.md:8]: will do, thanks!

**Review by irajdeep — APPROVED** [review]
LGTM 👍
